### PR TITLE
feat: add option to delete file to Recycle Bin

### PR DESCRIPTION
**Pull request by dyphire**
ref: https://github.com/tomasklaen/uosc/issues/390

### CODE REVIEW

**Comment by christoph-heinrich** [comment]
With this code being very similar to [delete-current-file.lua](https://github.com/stax76/mpv-scripts/blame/main/delete-current-file.lua#L59-L68) and GPL-2 technically not being compatible with GPL-3, we should probably ask for permission to make sure we're not stepping on anyone's toes.

@stax76 Are you fine with this?

**Comment by dyphire** [comment]
> With this code being very similar to [delete-current-file.lua](https://github.com/stax76/mpv-scripts/blame/main/delete-current-file.lua#L59-L68) and GPL-2 technically not being compatible with GPL-3, we should probably ask for permission to make sure we're not stepping on anyone's toes.

It's my mistake, @stax76 Can you tell me your opinion?

**Comment by tomasklaen** [comment]
1. Rename the option to `use_trash` so it's more universal, as only windows calls it a recycle bin.
2. I don't see any condition branch for macOS, so it's going to use `{ 'trash', path }`. It doesn't seem like mac has a `trash` command.
3. If there are some tools/packages required to be installed, it should be documented in the config file. But I'd be surprised if this couldn't be done without that. Can't we just `mv` files to some standard trash directory on linux/mac?
4. If the option is enabled, it should never fall back to deletion. If we are trying to delete on a platform where we can't implement it, it should log error message instead.

And make the condition branching easier to follow. So instead of:

```lua
if win and use_trash then
elseif win then
elseif not win and ...
```

do:

```lua
if win then
    if use_trash then
        -- trashing
    else
        -- deleting
    end
else if mac then
    -- mac if it differs from linux
else
    -- linux
end
```

**Comment by christoph-heinrich** [comment]
> 3\. If there are some tools/packages required to be installed, it should be documented in the config file. But I'd be surprised if this couldn't be done without that. Can't we just `mv` files to some standard trash directory on linux/mac?

`mv` doesn't create the necessary metadata file. We could create it ourselves, the format looks pretty simple.

**Comment by stax76** [comment]
It's fine, no problem.

**Comment by dyphire** [comment]
> 3\. If there are some tools/packages required to be installed, it should be documented in the config file. But I'd be surprised if this couldn't be done without that. Can't we just `mv` files to some standard trash directory on linux/mac?

I'm not familiar with linux and macos, so I'm not sure how to implement it.

**Comment by christoph-heinrich** [comment]
The spec for linux can be found here: https://specifications.freedesktop.org/trash-spec/trashspec-latest.html.
Before moving we'd have to also check for potential file conflicts in trash in addition to creating the metadata file. It's certainly easier to rely on trash-cli instead of implementing it ourselves.

**Comment by tomasklaen** [comment]
All right, 3rd party it is. So we need to find the macos alternative to `trash-cli`, use it in its branch, and document it in default config file.

Or if there is no alternative (I'd be surprised), document that the option is not available on macos.

**Comment by dyphire** [comment]
> All right, 3rd party it is. So we need to find the macos alternative to `trash-cli`, use it in its branch, and document it in default config file.
> 
> Or if there is no alternative (I'd be surprised), document that the option is not available on macos.

In fact, I found that there are third-party commands of `trash` on the macos. ref https://formulae.brew.sh/formula/trash

Done.

File: script-opts/uosc.conf [hunk 1]
```diff
@@ -183,6 +183,9 @@ media_types=3g2,3gp,aac,aiff,ape,apng,asf,au,avi,avif,bmp,dsf,f4v,flac,flv,gif,h
 subtitle_types=aqt,ass,gsub,idx,jss,lrc,mks,pgs,pjs,psb,rt,slt,smi,sub,sup,srt,ssa,ssf,ttxt,txt,usf,vt,vtt
 # Default open-file menu directory
 default_directory=~/
+# Move to the trash(recycle bin) when deleting file
+# On Linux and Macos the app trash-cli/trash must be installed first.
```
Comment (tomasklaen) [inline script-opts/uosc.conf:187]: Change the comment to:

```
# Move files to trash (recycle bin) when deleting files. Dependencies:
# - Linux: `sudo apt install trash-cli`
# - MacOS: `brew install trash`
```

File: scripts/uosc_shared/lib/utils.lua [hunk 2]
```diff
@@ -380,7 +380,29 @@ end
...
+
+            local escaped_path = string.gsub(path, "'", "''")
+            escaped_path = string.gsub(escaped_path, "’", "’’")
+            escaped_path = string.gsub(escaped_path, "%%", "%%%%")
+            ps_code = string.gsub(ps_code, "__path__", escaped_path)
+		    args = { 'powershell', '-NoProfile', '-Command', ps_code }
+		else
+			args = { 'cmd', '/C', 'del', path }
+		end
+	elseif state.os ~= 'windows' then
```
Comment (tomasklaen) [inline scripts/uosc_shared/lib/utils.lua:398]: You don't need `elseif` here. This is just an `else`.

Comment (dyphire) [reply scripts/uosc_shared/lib/utils.lua]: Done

File: script-opts/uosc.conf (hunk 1 above)
Comment (dyphire) [reply script-opts/uosc.conf]: Done

**Comment by tomasklaen** [comment]
If you enable `use_trash` in your `script-opts/uosc.conf`, all deletion done by uosc will go to trash.

On linux/mac this depends on trash-cli/trash being installed.

**Pull request merged**